python/cuml/cuml/neighbors/kneighbors_classifier.pyx

```diff
@@ -281,35 +350,46 @@ class KNeighborsClassifier(ClassifierMixin,
             n_classes = [len(c) for c in self._classes]
             ys = [self._y[:, i] for i in range(self._y.shape[1])]
 
-        probas = [
-            CumlArray.zeros(
-                (indices.shape[0], n), dtype=np.float32, order="C", index=indices.index
+        inds_ctype = <int64_t*><uintptr_t>inds.ptr
+
+        # Store Python attributes before nogil context
+        cdef size_t n_samples_fit = <size_t>self.n_samples_fit_
+        cdef size_t n_rows_size = <size_t>n_rows
+        cdef int n_neighbors_val = <int>self.n_neighbors
+
+        probas = []
+        for n, y in zip(n_classes, ys):
+            proba = CumlArray.zeros(
+                (n_rows, n), dtype=np.float32, order="C", index=inds.index
             )
-            for n in n_classes
-        ]
+            probas.append(proba)
+            proba_ptr = <float*><uintptr_t>proba.ptr
+            out_vec.push_back(proba_ptr)
 
-        cdef vector[int*] *y_vec = new vector[int*]()
-        cdef vector[float*] *proba_vec = new vector[float*]()
-        for proba, y in zip(probas, ys):
-            proba_vec.push_back(<float*><uintptr_t>proba.ptr)
+        cdef vector[int*] y_vec
+        for y in ys:
             y_vec.push_back(<int*><uintptr_t>y.ptr)
 
-        cdef handle_t* handle_ = <handle_t*><size_t>self.handle.getHandle()
-        cdef int64_t* indices_ptr = <int64_t*><uintptr_t>indices.ptr
-        cdef size_t n_samples = self._y.shape[0]
-        cdef size_t n_rows = indices.shape[0]
-        cdef int n_neighbors = self.n_neighbors
+        handle_ = <handle_t*><size_t>self.handle.getHandle()
+
+        # Compute weights if needed (nullptr for uniform weights)
+        cdef float* weights_ctype = <float*>0  # nullptr
+        weights_cuml = None
+        if self.weights not in (None, 'uniform'):
+            weights_cp = compute_weights(cp.asarray(dists), self.weights)
+            weights_cuml = CumlArray(weights_cp)
+            weights_ctype = <float*><uintptr_t>weights_cuml.ptr
```

Comment:
- No need to define `weights_cuml`
- No need to use `CumlArray` at all
- No need to convert `dists` here, `compute_weights` can handle that already

```suggestion
        if self.weights not in (None, 'uniform'):
            weights = compute_weights(dists, self.weights)
            weights_ctype = <float*><uintptr_t>weights.data.ptr
```

Same goes for everywhere this is called.